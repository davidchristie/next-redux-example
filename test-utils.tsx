import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { render } from "@testing-library/react";
import { exampleInitialState } from "./reducer";

export const createWrapper = initialState => {
  const reducer = (state = initialState) => state;
  const store = createStore(reducer);
  const Wrapper: React.FunctionComponent = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return { Wrapper };
};

export const renderConnectedComponent = (
  element: React.ReactElement,
  { initialState = exampleInitialState } = {}
) => {
  const { Wrapper } = createWrapper(initialState);
  const result = render(element, {
    wrapper: Wrapper
  });
  return { result };
};
