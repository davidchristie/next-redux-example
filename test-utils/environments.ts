const environments = ["development", "production", "test"];

export const describeAllEnvironments = (fn: (environment: string) => void) => {
  const originalEnv = process.env;

  beforeEach(() => {
    jest.resetModules();
    const mockEnv = { ...originalEnv };
    delete mockEnv.NODE_ENV;
    process.env = mockEnv;
  });

  afterEach(() => {
    process.env = originalEnv;
  });

  describe.each(environments.map(environment => [environment]))(
    'In Environment: "%s"',
    environment => {
      (process.env as any).NODE_ENV = environment;

      fn(environment);
    }
  );
};
