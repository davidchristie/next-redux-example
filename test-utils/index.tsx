export * from "./components";
export * from "./environments";
export * from "./objects";
export * from "./reducers";
export * from "./stores";
