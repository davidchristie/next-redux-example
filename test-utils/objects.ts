export const itShouldHaveMethod = <T>(instance: T, method: keyof T) => {
  it(`Should have ${method} method`, () => {
    expect(instance[method]).toBeInstanceOf(Function);
  });
};
