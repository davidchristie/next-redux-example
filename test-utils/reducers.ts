import { Reducer, AnyAction } from "redux";

export const itShouldHaveNextState = <S extends {}>(
  name: string,
  {
    action,
    expectedNextState,
    previousState,
    reducer
  }: {
    action: AnyAction;
    expectedNextState: S;
    previousState: S;
    reducer: Reducer<S>;
  }
) => {
  it(name, () => {
    const actualNextState = reducer(previousState, action);
    expect(actualNextState).toEqual(expectedNextState);
  });
};
