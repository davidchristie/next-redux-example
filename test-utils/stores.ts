import { Dispatch, Store } from "redux";

export const mockDispatch = (store: Store): jest.Mock<Dispatch> => {
  const dispatch = jest.fn(store.dispatch);
  store.dispatch = dispatch;
  return dispatch;
};
