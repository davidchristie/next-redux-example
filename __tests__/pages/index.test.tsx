import React from "react";
import App from "../../pages/index";
import { renderConnectedComponent } from "../../test-utils";

describe("index page", () => {
  describe("With React Testing Library", () => {
    it('Shows "Index Page"', () => {
      const { result } = renderConnectedComponent(<App />);
      expect(result.getByText("Index Page")).not.toBeNull();
    });
  });

  describe("With React Testing Library Snapshot", () => {
    it("Should match Snapshot", () => {
      const { result } = renderConnectedComponent(<App />);
      expect(result.asFragment()).toMatchSnapshot();
    });
  });
});
