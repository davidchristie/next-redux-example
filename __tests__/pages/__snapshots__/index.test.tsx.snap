// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index page With React Testing Library Snapshot Should match Snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Index Page
    </h1>
    <div
      class="jsx-1453841751 "
    >
      00:00:00
    </div>
    <div
      class="jsx-2435724200"
    >
      <h1
        class="jsx-2435724200"
      >
        Count: 
        <span
          class="jsx-2435724200"
        >
          0
        </span>
      </h1>
      <button
        class="jsx-2435724200"
      >
        +1
      </button>
      <button
        class="jsx-2435724200"
      >
        -1
      </button>
      <button
        class="jsx-2435724200"
      >
        Reset
      </button>
    </div>
    <nav>
      <a
        href="/other"
      >
        Navigate: Other Page
      </a>
    </nav>
  </div>
</DocumentFragment>
`;
